src/connection.ts

```diff
@@ -1,71 +1,197 @@
 
-	// Remove vscode-languageserver's stream listeners that kill the process if the stream is closed
-	input.removeAllListeners('end');
-	input.removeAllListeners('close');
-	output.removeAllListeners('end');
-	output.removeAllListeners('close');
+	constructor(input: NodeJS.ReadableStream) {
+		super();
+		const reader = new StreamMessageReader(input);
+		reader.listen(msg => {
+			this.emit('message', msg);
+		});
+		reader.onError(err => {
+			this.emit('error', err);
+		});
+		reader.onClose(() => {
+			this.emit('close');
+		});
+		this.setMaxListeners(Infinity);
+	}
 
-	return connection;
+	on(event: 'message', listener: (message: Message) => void): this;
+	on(event: 'error', listener: (error: Error) => void): this;
+	on(event: 'close', listener: () => void): this;
+	on(event: string, listener: Function): this {
+		return super.on(event, listener);
+	}
 }
 
 /**
  * Registers all method implementations of a LanguageHandler on a connection
+ *
+ * @param messageEmitter MessageEmitter to listen on
+ * @param messageWriter MessageWriter to write to
+ * @param handler TypeScriptService object that contains methods for all methods to be handled
+ * @param logger A logger that all messages will be logged to
+ * @param tracer An opentracing-compatible tracer
  */
-export function registerLanguageHandler(connection: IConnection, handler: LanguageHandler): void {
-
-	connection.onInitialize((params, token) => handler.initialize(params, token));
-	connection.onShutdown(() => handler.shutdown());
-
-	// textDocument
-	connection.onDidOpenTextDocument(params => handler.didOpen(params));
-	connection.onDidChangeTextDocument(params => handler.didChange(params));
-	connection.onDidSaveTextDocument(params => handler.didSave(params));
-	connection.onDidCloseTextDocument(params => handler.didClose(params));
-	connection.onReferences((params, token) => handler.getReferences(params, token));
-	connection.onHover((params, token) => handler.getHover(params, token));
-	connection.onDefinition((params, token) => handler.getDefinition(params, token));
-	connection.onDocumentSymbol((params, token) => handler.getDocumentSymbol(params, token));
-	connection.onCompletion((params, token) => handler.getCompletions(params, token));
-	connection.onRequest('textDocument/xdefinition', (params, token) => handler.getXdefinition(params, token));
-
-	// workspace
-	connection.onWorkspaceSymbol((params, token) => handler.getWorkspaceSymbols(params, token));
-	connection.onRequest('workspace/xreferences', (params, token) => handler.getWorkspaceReference(params, token));
-	connection.onRequest('workspace/xpackages', (params, token) => handler.getPackages(params, token));
-	connection.onRequest('workspace/xdependencies', (params, token) => handler.getDependencies(params, token));
+export function registerLanguageHandler(
+	messageEmitter: MessageEmitter,
+	messageWriter: MessageWriter,
+	handler: TypeScriptService,
+	logger: Logger = new NoopLogger(),
+	tracer = new Tracer()
```

Comment:
I initially had the logger only for logging the messages so you could just pass Noop for that. Will add option